db/silkworm/etl/file_provider.cpp

```diff
@@ -0,0 +1,66 @@
+/*
+   Copyright 2020 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "file_provider.hpp"
+
+namespace silkworm::etl{
+
+FileProvider::FileProvider(int id) {
+    filename_ = "tmp" + std::to_string(id);
+    file_.open(filename_, std::ios_base::in | std::ios_base::out | std::ios_base::trunc);
+}
+
+void FileProvider::write_buffer_to_disk(Buffer *buffer) {
+    union head_t {
+        uint32_t lengths[2];
+        uint8_t bytes[8];
+    };
+    head_t head{};
+    for(const auto& entry: buffer->get_entries()) {
+        head.lengths[0] = entry.key.size();
+        head.lengths[1] = entry.value.size();
+        file_.write((const char *) head.bytes, 8);
+        file_.write((const char *) entry.key.data(), entry.key.size());
+        file_.write((const char *) entry.value.data(), entry.value.size());
+    }
+    file_.seekg(0);
```

Comment:
Are all the writes correctly flushed to disk ? (check for any error like for exampe disk full)
You actually do not know if this instance of FileProvider is valid